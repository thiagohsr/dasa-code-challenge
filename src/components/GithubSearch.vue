<template>
  <form @submit.prevent="getRepositories" v-bind:class="classes.searchBox">
    <custom-input placeholder="Usuário do github" v-model="username" />
    <custom-button v-bind:disabled="!username">buscar</custom-button>
  </form>
</template>
<script>
import CustomButton from "@/components/CustomButton.vue";
import CustomInput from "@/components/CustomInput.vue";
import { getRepositories } from "@/services/github";
import { GITHUB_USER_REPOS } from "@/constants/urls";
import classes from "@/utils/cssTranspilation";
import parseLinkHeader from "@/utils/githubPaginationParser";

const styles = {
  searchBox: {
    display: "inline-flex",
    justifyContent: "center"
  }
};
export default {
  name: "GithubSearch",
  data() {
    return {
      classes: classes(styles),
      username: ""
    };
  },
  methods: {
    closeLoadingAnimation() {
      let animation = setInterval(() => {
        this.$store.commit("isLoading", false);
        clearInterval(animation);
      }, 900);
    },
    async getRepositories() {
      const url = GITHUB_USER_REPOS.replace("{{username}}", this.username);
      this.$store.commit("isLoading", true);
      this.$store.dispatch("getRepositories");
      const response = await getRepositories(url);

      if (response.message) {
        this.$store.commit("errorMessage", "O usuário informado não existe");
        this.closeLoadingAnimation();
        return;
      }
      if (!response.data.length) {
        this.$store.commit(
          "errorMessage",
          "Não foram encontrados repositórios para o usuário informado."
        );
        this.$store.commit("userRepositories", []);
        this.$store.commit("paginationLinks", {});
        this.closeLoadingAnimation();
        return;
      } else {
        this.$store.commit("errorMessage", "");
      }
      this.$store.commit("githubUser", this.username);
      this.$store.commit("userRepositories", response.data);

      if (response.headers.link) {
        this.$store.commit(
          "paginationLinks",
          parseLinkHeader(response.headers.link)
        );
      } else {
        this.$store.commit("paginationLinks", {});
      }
      this.closeLoadingAnimation();
    }
  },
  components: {
    "custom-button": CustomButton,
    "custom-input": CustomInput
  }
};
</script>
