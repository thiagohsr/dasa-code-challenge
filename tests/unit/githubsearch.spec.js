import GithubSearch from "@/components/GithubSearch.vue";
import { createLocalVue, mount, shallowMount } from "@vue/test-utils";
import Vuex from "vuex";

const localVue = createLocalVue();
localVue.use(Vuex);

describe("GithubSearch component tests", () => {
  let actions;
  let store;
  let state;
  let mutations;
  let getters;
  beforeEach(() => {
    state = {
      githubUser: "",
      errorMessage: "",
      userRepositories: [],
      paginationLinks: {},
      isLoading: false
    };
    mutations = {
      errorMessage(state, errorMessage) {
        state.errorMessage = errorMessage;
      },
      githubUser(state, githubUser) {
        state.githubUser = githubUser;
      },
      userRepositories(state, userRepositories) {
        state.userRepositories = userRepositories;
      },
      paginationLinks(state, links) {
        state.paginationLinks = links;
      },
      isLoading(state, isLoading) {
        state.isLoading = isLoading;
      }
    };
    getters = {
      githubUser: state => {
        return state.githubUser;
      }
    };
    actions = {
      getRepositories: ({ commit }, { githubUser }) => {
        commit("githubUser", githubUser);
      },
      githubUser: ({ commit }, { githubUser }) => {
        commit("githubUser", githubUser);
      }
    };
    store = new Vuex.Store({
      state,
      actions,
      mutations,
      getters
    });
  });
  it("should render properly", () => {
    const shallowed = shallowMount(GithubSearch);
    expect(shallowed).toMatchSnapshot();
    expect(shallowed.name()).toBe("GithubSearch");
  });
  it("should render properly child components", () => {
    const shallowed = shallowMount(GithubSearch);
    const customInput = shallowed.find("custom-input-stub");
    const customButton = shallowed.find("custom-button-stub");
    expect(customInput.name()).toBe("CustomInput");
    expect(customButton.name()).toBe("CustomButton");
  });
  it("should enable button when fill input", () => {
    const mounted = mount(GithubSearch);
    const customInput = mounted.find("input");
    const customButton = mounted.find("button");
    expect(customButton.attributes().disabled).toBe("disabled");
    customInput.setValue("githubUser");
    expect(customButton.attributes().disabled).toBe(undefined);
  });

  it("should fill up username data with input value", async () => {
    const expectedValue = "githubuser";
    const mounted = mount(GithubSearch, { store, localVue });
    const customInput = mounted.find("input");
    customInput.setValue(expectedValue);
    expect(mounted.vm.$data.username).toBe(expectedValue);
  });

  it("should call getRepositories method", async () => {
    const mounted = mount(GithubSearch, { store, localVue });
    const customInput = mounted.find("input");
    const customButton = mounted.find("button");
    mounted.vm.getRepositories = jest.fn();
    customInput.setValue("githubUser");
    customButton.trigger("submit");
    expect(mounted.vm.getRepositories).toBeCalled();
  });

  it("should call getRepositories action from store", async () => {
    const mounted = mount(GithubSearch, {
      store,
      localVue
    });
    const expectedValue = "githubUser";

    const customInput = mounted.find("input");
    const customButton = mounted.find("button");
    customInput.setValue(expectedValue);
    customButton.trigger("submit");
    expect(mounted.vm.$store.getters.githubUser).toBe(expectedValue);
  });
});
