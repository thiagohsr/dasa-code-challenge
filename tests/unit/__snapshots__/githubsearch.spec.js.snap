// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GithubSearch component tests should render properly 1`] = `
<form class="searchBox-0-0-1">
  <custom-input-stub value="" placeholder="Usuário do github"></custom-input-stub>
  <custom-button-stub label="Botão" disabled="true">buscar</custom-button-stub>
</form>
`;
